[DO NOT SQUASH] Rebased and compressed version of #1468

pcf000 opened the pull request:
I couldn't figure out how to split pull request #1468 into two or more that would still run, and the recent main-line issues didn't help.  I rebased on develop to get past the conflicts from the upstream merge and I combined the commits into four of roughly useful contents.  Otherwise, it's the same as #1468.  I didn't change #1468 itself because I was using an experimental script for the rebase.

Do-not-squash to preserve the few separate commits.

pcf000 committed:
Changing 'rock.conv2d' to 'rock.conv'.

Behind-the-scenes changes, mostly related to ConvolutionDims.

Convert x/y to 1/0 in ConvToGemm.cpp, though still no loops, and some affected tests.

Make strideH/W into an array, first step to doing dilation, padding, 'tilda', etc.

Making arrays out of some dilation and stride variables, and dependents.

Making arrays out of padding variables and dependents.

Methodical test fixing.  Still fails the two resnet18 pr-e2e xmir tests.  Some debug code still present.

To fix the two resnet18_blk* tests, have both hw and 01 entries in the match-layouts-to-input tables.

Also needed to change conv2d to conv in tuning code.

Change hw to 01 for pad, tilda, slice.  Use 0f and 0i when both y and h are being used.

Finish converting scalars to vectors and loops in ConvToGemm.cpp
For expanding names, make strings that can be aggregated into a
StringMap for passing as an initialisation.  For most other things,
either loop doing the operation for the given dimension index, or loop
across all dimensions to accumulate a list of strings to pass to a
transform function.
All created strings are made with b.getStringAttr() so they stick around
across iterations.

pcf000 committed:
Big search-and-replace of y/x/h/w to 0/1, except for seven Tosa tests that'll come with #1314.

pcf000 committed:
Generalise backwardData()'s delinearisation of kernelId, though only 3-D is added.  Small cleanups.

Convert getGemmSize() and backwardDataKernelIds() to use loops over N dimensions.

Fix bug in getKernelCount() array version.

Convert hasValidDimension() to use dimension arrays, and some more concise llvm functions.

Make a loop to add 0,1,etc key names in maps for MatchLayoutsToInput.

More assorted 2d-assumption fixes.  Not passing SparseTensor tests.

Handle 1, 2, or 3 dimensions in generateKernel().

Change parseConvDims() to take an array of dimensions instead of separate height/width/(depth).

Corrections after merge.  Fix self-inflicted bug in getConvolutionDims().  Passes check-rocmlir now.

pcf000 committed:
Getting a rock.conv3d test case to compile.  Debug code and crude ungeneralised code still present.

Flush debug code, add a self-verifying 3d-rock.conv test.

Fix error in changes to MIGraphXToTosa.cpp.

Fix another error in changes to MIGraphXToTosa.cpp.

Account for both 2- and 3-D convolutions in isValid() key check.

Changing ConvolutionDims.h to avoid clang-tidy problems.

Refactor and generalise correctConvParameters().

Rebase via merge. 'pf-issue-1311e' rebased on 'origin/develop'.

Some renamings for changes that happened in the meantime.

krzysz00 commented on mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:
```diff
@@ -214,52 +214,82 @@ LogicalResult ConvConverter<ConvType>::matchAndRewrite(
   SmallVector<int64_t> NCHW2NHWC{0, 2, 3, 1};
   SmallVector<int64_t> NHWC2NCHW{0, 3, 1, 2};
 
+  int dims = outputTy.getShape().size() - 2;
```
Nit: `size_t`? Or `uint32_t`?

pcf000 committed:
Change a couple of program invocations to use /home/pf/bin:/home/pf/.local/bin:/usr/local/bin:/home/pf/bin:/home/pf/.local/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin instead of /opt/rocm/llvm/bin,
because rocm 6.1 doesn't include git-clang-format or llvm-profdata in the rocm-llvm
package but rather in the optional rocm-llvm-dev package, and our docker image lacks them.
